torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,536 @@
+// Proof: If b is a multiple of c, then we have b % c = 0
+// Also, according to Theorem 4.13, (ac) % b = (a % (b / c)) * c
+// From the fundamental division-with-remainder equation, we have:
+// ac = ((ac)/b)b + (ac)%b ... (eq 1)
+// a = (a/(b/c))(b/c) + a%(b/c) ... (eq 2)
+// Multiply c to both side of (eq 2), and note that b is a multiple of c:
+// ac = (a/(b/c))b + (a%(b/c)) * c
+// subtrace the above equation with (eq 1), we get
+// 0 = [a/(b/c) - (ac)/b] * b + [(a%(b/c))*c - (ac)%b]
+// The second term in the above equation is 0, so we have
+// 0 = [a/(b/c) - (ac)/b] * b
+// because b != 0, we get a/(b/c) = (ac)/b
+//
+// Theorem 4.15: If a % c + b % c < |c|, then (a+b)/c = a/c + b/c
+// Proof: From Theorem 4.7 and Theorem 4.5, we have
+// (a + b) % c = (a % c + b % c) % c = a % c + b % c
+// From the fundamental division-with-remainder equation, we have:
+// (a + b) = ((a + b) / c) * c + (a + b) % c
+// a = (a / c) * c + a % c
+// b = (b / c) * c + b % c
+// Adding the last two equations and subtract with the first equation:
+// 0 = [(a/c + b/c) - (a+b)/c] * c + [a % c + b % c - (a + b) % c]
+// The second term in the above equation is 0, so we have
+// 0 = [(a/c + b/c) - (a+b)/c] * c
+// Because c != 0, we have (a/c + b/c) = (a+b)/c
+//
+// 5. Implementations of Div and Mod
+// =================================
+//
+// Unfortunately, modern hardwares and programming languages does not implement
+// div and mod consistent with Euclid's division lemma, although these
+// implementations can be converted with Euclid's division easily. The
+// implementations of div and mod depends on programming languages. The
+// comparison of these implementations and their properties are discussed in the
+// following paper:
+//
+// Boute, Raymond T. "The Euclidean definition of the functions div and mod."
+// ACM Transactions on Programming Languages and Systems (TOPLAS) 14.2 (1992):
+// 127-144.
+//
+// I will summarize some useful points from the above paper here, and add my own
+// comments:
+//
+// For a >= 0 and b > 0, all implementation of a/b and a%b are the same and
+// consistent with the Euclid's division. So no brainer in this region.
+//
+// Except for a few Languages (ISO Standard Pascal, Algol, Ada) all
+// implementations are consistent with the fundamental division-with-remainder
+// equation, although the range and sign of r can be different and the value of
+// q can be different by 1. Implementations not satisfying the fundamental
+// division-with-remainder equation is considered wrong because it has no
+// mathematical properties. For all implementations, |a % b| < |b|.
+//
+// Common implementations are:
+//
+// trunc div (round to zero):
+// a/b := trunc(a \ b)
+// a%b defined by the fundamental division-with-remainder equation
+//
+// floor div:
+// a/b := floor(a \ b)
+// a%b defined by the fundamental division-with-remainder equation
+//
+// For C89, the result of negative div is not specified. C99 and C++ uses trunc
```

Comment:
Might be helpful to cite the C99 standard's discussion of truncation division

Reply:
added to section "6. Properties of Div and Mod Under Trunc Div"